cypress/integration/unit_tests/polygon_draw_test.js

```diff
@@ -49,142 +67,182 @@ describe('Unit test for ShapeData', () => {
+    const confirmStub = cy.stub(window, 'confirm').returns(true);
+    drawPolygonAndClickOnIt();
+    pressPopupButton('edit').then(() => getFirstFeature().setPath(newPath));
+    pressButtonAndWaitForPromise('close').then(() => {
+      expect(confirmStub).to.be.calledOnce;
+      expect(convertPathToLatLng(getFirstFeature().getPath())).to.eql(path);
     });
+    assertOnFirestoreAndPopup(path, defaultData);
   });
 
-  it('Update while update pending', () => {
-    const popup = new StubPopup();
-    popup.setCalculatedData(calculatedData);
-    const underTest =
-        new StoredShapeData('my_id', 'my notes', polygonGeometry, popup);
-    const records = [];
-    const ids = [];
-    const setThatTriggersNewUpdate = {
-      set: (record) => {
-        popup.notes = 'racing notes';
-        expect(underTest.update()).to.be.null;
-        expect(popup.calculatedData).to.eql(calculatedData);
-        expect(underTest.state).to.eql(StoredShapeData.State.QUEUED_WRITE);
-        records.push(record);
+  it('Updates while update pending', () => {
+    let fakeCalled = false;
+    drawPolygonAndClickOnIt().then(() => {
+      currentUpdatePromise = null;
+      const [, data] = getFirstUserRegionDataEntry();
+      const realDoc = userShapes.doc(data.id);
+      const realDocFunction = userShapes.doc;
+      const fakeDoc = {};
+      userShapes.doc = () => fakeDoc;
+      fakeDoc.set = (record) => {
+        console.log('not sure how to handle this part');
+        // Unfortunately Cypress can't really handle executing Cypress commands
+        // inside an asynchronous callback, so we rely on jquery to modify the
+        // DOM. Works out ok.
+        $('button:contains("edit")').trigger('click');
+        $('.notes').val('racing notes');
+        $('button:contains("save")').trigger('click');
+        expect(currentUpdatePromise).to.be.null;
+        expect(data.state).to.eql(StoredShapeData.State.QUEUED_WRITE);
         expect(StoredShapeData.pendingWriteCount).to.eql(1);
-        // Put back usual set for next run.
-        setThatTriggersNewUpdate.set = recordRecord(records, null);
-        return Promise.resolve(null);
-      },
-    };
-    firebaseCollection.doc = (id) => {
-      ids.push(id);
-      return setThatTriggersNewUpdate;
-    };
-    popup.notes = 'new notes';
-    cy.wrap(underTest.update()).then(() => {
-      expect(ids).to.eql(['my_id', 'my_id']);
-      expect(records).to.have.length(2);
-      expect(records).to.eql([
-        {
-          calculatedData: calculatedData,
-          geometry: polygonGeometry,
-          notes: 'new notes',
-        },
-        {
-          calculatedData: calculatedData,
-          geometry: polygonGeometry,
-          notes: 'racing notes',
-        },
-      ]);
-      expect(underTest.id).to.eql('my_id');
-      expect(StoredShapeData.pendingWriteCount).to.eql(0);
+        userShapes.doc = realDocFunction;
+        fakeCalled = true;
+        return realDoc.set(record);
```

Comment:
Why can't you just stub out readDoc.set (store the function and call it at the end?)

Reply:
Discussed offline, and added a comment.